### DISCUSSION TITLE
Ruler consistency issue during store gateway deployment

### DISCUSSION THREAD

**Discussion by callumj** [grafana/mimir#4092] [Help and support]
We performed a deployment of our Mimir environment (2.5.0) that involved redeploying all components (SGW, Ruler, etc) and during this deployment our Ruler observed some issues querying from the SGW:

```
	the consistency check failed because some blocks were not queried (err-mimir-store-consistency-check-failed). The non-queried blocks are: 01GP1J6W4BXRW0CARPCC45M51V
```

After ~30 minutes this error went away and the rules were being evaluated without errors.

We verified that SGWs were healthy (other than a new set starting up and a old set slowly scaling down as each new pod came up). We checked the SGW side and we couldn't find any error messages referring to those blocks which leaves us a little stumped. 

One theory we have is maybe the Ruler was querying recently started SGWs that may have not loaded S3 metadata in to memory yet, and in this case maybe we haven't configured SGW/Ruler correctly (e.g. are we missing a flag where SGWs are being registered too early?).

Some configuration that might be helpful:
* `query_store_after: 24h` (so we query recent data from ingesters)
* `query_ingesters_within: 25h`
* `ignore_blocks_within: <default>`

**Comment by pracucci** [comment]
I think your theory is valid. This error typically happens when multiple store-gateways are restarting at the same time. When running Mimir with the default configure (store-gateway replication factor = 3), you should make sure that:

- If you deploy store-gateways **without** zone-aware replication: no more than 2 store-gateways restart at the same time
- If you deploy store-gateways **with** zone-aware replication: no more than 2 store-gateway zones restart at the same time

The store-gateway readiness probe succeeds only after their blocks have been synched and they're ready to serve queries.
